zkp-prover/src/zkp_prover.rs

```diff
@@ -0,0 +1,161 @@
+    ) -> Self {
+        let network_info = NetworkInfo::from_network_id(blockchain.read().network_id());
+        let genesis_block = network_info.genesis_block::<Block>().unwrap_macro();
+
+        let public_keys = zkp_state.read().latest_pks.clone();
+        let genesis_state = state_commitment(
+            genesis_block.block_number(),
+            genesis_block.hash().into(),
+            public_keys,
+        );
+
+        let blockchain_election_rx = blockchain.write().notifier.as_stream();
+
+        Self {
+            blockchain,
+            network,
+            zkp_state,
+            genesis_state,
+            receiver: None,
+            blockchain_election_rx,
+            proof_generation_thread: None,
+        }
+    }
+
+    pub fn is_generating_proof(&self) -> bool {
+        self.proof_generation_thread.is_some()
+    }
+
+    fn broadcast_zk_proof(network: &Arc<N>, zk_proof: ZKProof) {
+        let network = Arc::clone(network);
+        tokio::spawn(async move {
+            if let Err(e) = network.publish::<ZKProofTopic>(zk_proof).await {
+                log::warn!(error = &e as &dyn Error, "Failed to publish the zk proof");
+            }
+        });
+    }
+
+    pub(crate) fn cancel_current_proof_production(&mut self) {
+        self.receiver = None;
+        if let Some(ref thread) = self.proof_generation_thread {
+            thread.abort();
+            self.proof_generation_thread = None;
+        }
+    }
+}
+
+impl<N: Network> Stream for ZKProver<N> {
+    type Item = ZKProof;
+
+    fn poll_next(mut self: Pin<&mut ZKProver<N>>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        // If a new proof was generated it sets the state and broadcasts the new proof
+        // This returns early to avoid multiple spawns of proof generation
+        if let Some(ref mut receiver) = self.receiver {
+            return match receiver.poll_unpin(cx) {
+                Poll::Ready(Ok(Ok(new_zkp_state))) => {
+                    assert!(
+                        new_zkp_state.latest_proof.is_some(),
+                        "The generate new proof should never produces a empty proof"
+                    );
+                    self.receiver = None;
+                    let zkp_state_lock = self.zkp_state.upgradable_read();
+                    if zkp_state_lock.latest_block_number < new_zkp_state.latest_block_number {
+                        let mut zkp_state_lock = RwLockUpgradableReadGuard::upgrade(zkp_state_lock);
+                        *zkp_state_lock = new_zkp_state;
+
+                        let zkp_state_lock = RwLockWriteGuard::downgrade(zkp_state_lock);
+
+                        let proof: ZKProof = zkp_state_lock.clone().into();
+                        Self::broadcast_zk_proof(&self.network, proof.clone());
+                        return Poll::Ready(Some(proof));
+                    }
+                    Poll::Pending //ITODO Doesn't need to return early but it's cleared to do so (?)
```

Comment:
Actually, we might need a timer in the zkp component, just like in the consensus.